isisd: Segment Routing support

rwestphal opened the pull request:
This is an alternate implementation of SR based on Olivier's code (#4800). Opening it for review as requested on the meeting today.

When reviewing, please fetch the PR locally in order to see the commits in their original order.

odd22 commented on isisd/isis_sr.c:
```diff
@@ -0,0 +1,1654 @@
+		exit(1);
+	}
+}
+
```
IPv4 and IPv6 delete functions could be merge with a new `family` parameter to separate specific IPv4 from IPv6 stuff.

rwestphal replied:
Agreed. I'll merge the IPv4/IPv6 Adj-SID functions to reduce code duplication. [2]

odd22 commented on isisd/isis_sr.c:
```diff
@@ -0,0 +1,1654 @@
+	isis_sr_adj_sid_install_uninstall(true, circuit, adj, local_label,
+					  AF_INET6, &adj->ipv6_addresses[0]);
+}
+
```
IPv4 and IPv6 add functions could be merge with a new `family` parameter to separate specific IPv4 from IPv6 stuff.

rwestphal replied:
Agreed. I'll merge the IPv4/IPv6 Adj-SID functions to reduce code duplication.

odd22 commented on isisd/isis_sr.c:
```diff
@@ -0,0 +1,1654 @@
+
+	return 0;
+}
+
```
Same. IPv4 & IPv6 functions could be merge

rwestphal replied:
Agreed. I'll merge the IPv4/IPv6 Adj-SID functions to reduce code duplication. [3]

odd22 commented on isisd/isis_sr.c:
```diff
@@ -0,0 +1,1654 @@
+		    && !CHECK_FLAG(pcfg->flags, ISIS_PREFIX_SID_NODE)) {
+			SET_FLAG(pcfg->flags, ISIS_PREFIX_SID_NODE);
+			lsp_regenerate_schedule(area, area->is_type, 0);
+		}
```
Node SID flags is not correctly report in PSID.

rwestphal replied:
Good catch. I fixed the issue and now the N-flag is being set appropriately for routes associated to loopback interfaces [3].

odd22 commented on isisd/isis_sr.c:
```diff
@@ -0,0 +1,1654 @@
+
+	return CMD_SUCCESS;
+}
+
```
New show commands print only prefix information disregarding the advertised SR-Node. SR-Node information, in particular SRGB is essential for debugging especially to detect SRGB overlap, index out of SRGB ...

rwestphal replied:
Since the table already shows the SID and input/output labels of each nexthop, I don't think the full SRGB information is necessary there. I'd prefer to not add any new column to the table to prevent it from getting too wide. The operators will always have `show isis database detail` as a last resort if they need full detailed information.

odd22 commented on isisd/isis_tlvs.c:
```diff
@@ -3467,10 +4387,22 @@ static void tlvs_protocols_supported_to_adj(struct isis_tlvs *tlvs,
+	else if (adj->ipv4_address_count > 0 && tlvs->ipv4_address.count == 0)
+		hook_call(isis_adj_ipv4_disabled_hook, adj);
+
 	if (adj->ipv4_address_count != tlvs->ipv4_address.count) {
```
The `else if` is never executed when adjacency is removed due to a bug in isisd code. 

If `isis_adj_ipv4_XXX_hook()` is merged with IPv6 one, it is easier to add AF_INET parameter here.

Why not using a direct call to `isis_sr_adj_ipv4_enabled()` function instead of using hook_call() ?

rwestphal replied:
This hook will be called only when the adjacency runs out of IPv4 addresses. When the adjacency goes down altogether, the `isis_adj_state_change_hook` hook is called instead. The SR code registers a callback to both hooks to handle these different events.

> Why not using a direct call to `isis_sr_adj_ipv4_enabled()` function instead of using hook_call() ?

Mostly to make the code more modular. In the future we might want to register other callbacks to this hook.

rwestphal commented on isisd/isis_sr.c:
```diff
@@ -0,0 +1,1654 @@
+	if (ifp && isis_sr_prefix_is_node_sid(ifp, prefix))
+		SET_FLAG(pcfg->flags, ISIS_PREFIX_SID_NODE);
+
+	/* Save prefix-sid configuration. */
```
Good catch. I fixed the issue and now the N-flag is being set appropriately for routes associated to loopback interfaces.

rwestphal commented on isisd/isis_sr.c:
```diff
@@ -0,0 +1,1654 @@
+	}
+	if (external)
+		SET_FLAG(psid->flags, ISIS_PREFIX_SID_READVERTISED);
+
```
Good catch. I fixed the issue and now the N-flag is being set appropriately for routes associated to loopback interfaces [2].

rwestphal commented on isisd/isis_adjacency.h:
```diff
@@ -111,13 +122,18 @@ struct isis_adjacency *isis_adj_lookup(const uint8_t *sysid,
+DECLARE_HOOK(isis_adj_ipv4_disabled_hook, (struct isis_adjacency *adj), (adj))
+DECLARE_HOOK(isis_adj_ipv6_enabled_hook, (struct isis_adjacency *adj), (adj))
+DECLARE_HOOK(isis_adj_ipv6_disabled_hook, (struct isis_adjacency *adj), (adj))
 void isis_adj_state_change(struct isis_adjacency *adj,
```
That's good idea, I'll do that.

rwestphal commented on isisd/isis_sr.c:
```diff
@@ -0,0 +1,1654 @@
+	if (fragnode) {
+		lsp = listgetdata(fragnode);
+		goto lspfragloop;
+	}
```
I used the same idiom used by the SPF code to iterate over LSP fragments. But I agree a proper loop construct will make the code easier to follow. I'll change this.

rwestphal commented on isisd/isis_sr.c:
```diff
@@ -0,0 +1,1654 @@
+		if (IS_DEBUG_ISIS(DEBUG_SR))
+			zlog_debug("  no SR-capable nexthops");
+		return -1;
+	}
```
I think goto is fine for error handling like this, I don't see the need to make any change here.

rwestphal commented on isisd/isis_sr.c:
```diff
@@ -0,0 +1,1654 @@
+			  __func__, srp->sid.value);
+		return -1;
+	}
+
```
Fair enough. I'll split this function into two to accommodate this requirement.

rwestphal commented on isisd/isis_sr.c:
```diff
@@ -0,0 +1,1654 @@
+			__func__, &srp->prefix);
+		return -1;
+	}
+
```
Fair enough. I'll split this function into two to accommodate this requirement. [2]

rwestphal commented on isisd/isis_sr.c:
```diff
@@ -0,0 +1,1654 @@
+		/* Map absolute label to SID index. */
+		psid->value = pcfg->sid - srdb->config.srgb_lower_bound;
+		break;
+	}
```
1. It seems we have conflicting information here. My implementation of absolute SIDs is compatible with what IOS-XR does (as documented on Clarence's book) and also compliant to the SR draft according to one of its authors. I should start an email thread with you and other people to discuss this in more detail.
2. If you configure an absolute label that can't be mapped to the local SRGB, the error will be detected later (post SPF), the Prefix-SID will marked as invalid and an error will be logged accordingly.

rwestphal commented on isisd/isis_sr.h:
```diff
@@ -0,0 +1,223 @@
+	/* Backpointer to IS-IS area. */
+	struct isis_area *area;
+};
+
```
I think this is a matter of personal preference. For me, the distinction of configuration data from runtime data makes the code much easier to follow.

riw777 replied:
Keeping them separate also probably makes it easier to manage at scale.

rwestphal commented on isisd/isis_lsp.c:
```diff
@@ -908,22 +920,38 @@ static void lsp_build(struct isis_lsp *lsp, struct isis_area *area)
+			lsp->tlvs, id, area->srdb.config.srgb_lower_bound,
+			area->srdb.config.srgb_upper_bound,
+			area->srdb.config.msd);
+		lsp_debug("ISIS (%s): Adding Router Capabilities information",
```
Ok, I'll change the code to send the Router Capability TLV even when SR is not enabled.

rwestphal commented on isisd/isis_sr.h:
```diff
@@ -0,0 +1,223 @@
+		struct tree_sr_prefix_cfg_head prefix_sids;
+	} config;
+};
+
```
The `sr_prefix->type` member already specifies the type of the Prefix-SID (either `ISIS_SR_PREFIX_LOCAL` or `ISIS_SR_PREFIX_REMOTE`). This is used to give local Prefix-SIDs the special treatment they need.

rwestphal commented on isisd/isis_sr.h:
```diff
@@ -0,0 +1,223 @@
+#define F_ISIS_SR_NODE_NEW		0x01
+#define F_ISIS_SR_NODE_MODIFIED		0x02
+#define F_ISIS_SR_NODE_UNCHANGED	0x04
+};
```
I don't see why we'd need to do that any of that now. If we need to collect more information to implement whatever new feature in the future, that can be done as part of the development of that feature IMO.

riw777 replied:
Using prefix sids as "node sids" is probably the primary use case for data center fabrics, since you don't/shouldn't have any parallel links... It might be worth discussing, but focusing on prefix sids for now seems reasonable to me.

rwestphal replied:
@riw777 Adj-SIDs are supported by this code. Olivier was requesting a different way to store Adj-SIDs in order to facilitate the implementation of upcoming features like TI-LFA. This request was addressed in one of my previous updates.

rwestphal commented on isisd/isis_adjacency.h:
```diff
@@ -103,6 +104,16 @@ struct isis_adjacency {
+			struct isis_lan_adj_sid *ladj_sid;
+		} ipv6;
+	} sr;
 };
```
1. It should be easy to adapt the code to handle Backup Adj-SIDs when necessary in the future. For now we can ignore them for the sake of simplicity.
2. Persistent Adj-SIDs are not implemented yet (I think the LM should provide infrastructure for that first).
3. Yes, isisd is not releasing Adj-SID labels when the corresponding adjacencies go down. I'll fix this by borrowing some code from ldpd. My long term plan, however, is to change the LM to provide a higher level API that manages label chunks transparently to the user (this way we avoid code duplication across different daemons).

riw777 replied:
It would be better for the LM to manage label chunks, rather than managing them across many different parts of the code base.

rwestphal commented on isisd/isis_northbound.c:
```diff
@@ -1513,6 +1522,262 @@ static int isis_instance_mpls_te_router_address_destroy(enum nb_event event,
+	pcfg->sid = yang_dnode_get_uint32(dnode, NULL);
+
+	return NB_OK;
+}
```
Yes, the local Prefix-SIDs aren't validated at configuration time (similar to IOS-XR). I think logging SID overflows to the terminal is a reasonable thing to do, but I'm not sure if the command should be rejected in that case. I need to think more about this.

riw777 replied:
If we reject the command, would we not advertise and prefix sids at all? I suspect this is the correct answer, because -- we don't know which prefixes are "more important," and we want to make certain the user "sees" this error, and the behavior is consistent. Better for things to be totally broken than partially, I think, in this case.

One thing I might be missing here -- is it possible to use tags or some such to determine which prefixes have sids assigned? Is this worth doing?

rwestphal replied:
> If we reject the command, would we not advertise and prefix sids at all? I suspect this is the correct answer, because -- we don't know which prefixes are "more important," and we want to make certain the user "sees" this error, and the behavior is consistent. Better for things to be totally broken than partially, I think, in this case.

After my latest updates, isisd doesn't accept the configuration of invalid Prefix-SIDs anymore.

> One thing I might be missing here -- is it possible to use tags or some such to determine which prefixes have sids assigned? Is this worth doing?

There's a new `show isis segment-routing prefix-sids` command that should fit that purpose.

Example:
```
debian# show isis segment-routing prefix-sids 
Area 1:
 IS-IS L1 Prefix-SIDs:

 Prefix                SID  In Label  Nexthop                             Out Label  Uptime    
 ----------------------------------------------------------------------------------------------
 1.1.1.1/32            10   -         local                               -          -         
 2.2.2.2/32            20   16020     10.0.1.2, eth-sw1                   16020      00:00:50  
 3.3.3.3/32            30   16030     10.0.1.3, eth-sw1                   17030      00:00:50  
 4.4.4.4/32            40   16040     10.0.1.2, eth-sw1                   16040      00:00:50  
 5.5.5.5/32            50   16050     10.0.1.3, eth-sw1                   17050      00:00:50  
 6.6.6.6/32            60   16060     10.0.1.3, eth-sw1                   17060      00:00:50  
                                      10.0.1.2, eth-sw1                   16060      00:00:50  
 2001:db8:1000::1/128  11   -         local                               -          -         
 2001:db8:1000::2/128  21   16021     fe80::687b:b3ff:fe01:8df0, eth-sw1  16021      00:00:50  
 2001:db8:1000::3/128  31   16031     fe80::54d1:6aff:fed0:ded8, eth-sw1  17031      00:00:50  
 2001:db8:1000::4/128  41   16041     fe80::687b:b3ff:fe01:8df0, eth-sw1  16041      00:00:50  
 2001:db8:1000::5/128  51   16051     fe80::54d1:6aff:fed0:ded8, eth-sw1  17051      00:00:50  
 2001:db8:1000::6/128  61   16061     fe80::54d1:6aff:fed0:ded8, eth-sw1  17061      00:00:50  
                                      fe80::687b:b3ff:fe01:8df0, eth-sw1  16061      00:00:50
```

rwestphal commented:
Just pushed another update (rebase + further changes).

@odd22 thanks for the excellent in-depth review of the PR. I did my best to address all of your review comments, or at least most of them. I think the code is better prepared now to accommodate other features we're planning to work on later (e.g. TI-LFA, flex algo, LDP interop, etc).

Please see my inline comments below.

> There is some interoperability problems with the code when tested against Cisco and Juniper router. In particular:
> 
> * Prefix are not advertised with the Node SID flag leading to improper MPLS encapsulation of IP packet on Cisco and Juniper side

Fixed. Now the N-flag is being set appropriately for routes associated to loopback interfaces.

> * Absolute label for Prefix is marshmallow. It doesn't correspond to anything in the draft and not work like Cisco implement it. And it is not supported by Juniper router. Prefix SID is advertised even if label is invalid.

Being discussed over email at the SPRING mailing list.

> * In case of interface flapping, code could consume all Label space

The code now releases Adj-SID labels from the LM when necessary. You can take a look at the `isis_zebra_release_dynamic_label()` function if you want to do this in your branch as well.

> Implementation choice is not ready for future Segment Routing features. In particular:
> 
> * Adjacency SID is not store in SRDB while it is necessary for TI-LFA, R-LFA ans SR-TE

I added an rb-tree to store Adj-SIDs on a per-area basis. The rb-tree will allow us to perform fast Adj-SID lookups later when necessary.

I also added a new linked list to store Adj-SIDs on a per-adjacency basis, which is useful to optimize a few operations (e.g. check which Adj-SIDs need to be removed when an adjacency goes down).

> * Backup ADJ-SID is not supported while it is in OSPF-SR version and it is necessary for TI-LFA & R-LFA. With Cisco and Juniper, FRR node could not be used as backup router in certain topology

I'm not sure if it makes sense to advertise and install backup Adj-SIDs when we don't support TI-LFA yet, but I changed the code to do that nonetheless.

> * Prefix nexthop computation and MPLS enforcement is performed within the same function which render evolution difficult e.g. adding Strict SPF support, Flexible Algorithm ...

I split the Prefix-SID processing/installation into separate functions.

> Finally, the code is difficult to read and thus will be hard to maintain due to:
> 
> * Lack of comments, in particular there is no description of the new functions
> * There is no technical document explaining how the code is articulate and working
> * There is too few debug messages that made debugging difficult

I added a few more comments and debug messages in the code to improve maintainability and troubleshooting.

> * All new functions start with the `isis_sr` prefix disregarding there are static or not. `isis_sr`must be reserved for non static functions

It seems isisd loosely follows this pattern in a couple of files, but that's far from being a rule. I'd prefer to keep the current names to avoid collisions with functions from other daemons (that can be annoying when using ctags and cscope).

rwestphal commented:
New update:
* Fixed the checkpatch issues that made sense to fix;
* Added a new comprehensive topotest for SR: https://github.com/FRRouting/frr/pull/4945/commits/6cb33bc25eb63d4bd72314231fe0cac7591974b9

riw777 commented on isisd/isis_adjacency.c:
```diff
@@ -121,6 +122,50 @@ struct isis_adjacency *isis_adj_lookup_snpa(const uint8_t *ssnpa,
+			for (int level = IS_LEVEL_1; level <= IS_LEVEL_2;
+			     level++) {
+				if ((area->is_type & level) == 0)
+					continue;
```
I'm a little confused by this line -- what does this do?

rwestphal replied:
It's a common idiom used in the isisd code to handle L1/L2 areas like both L1 and L2 areas (which is what they are).

Please take a look below:
```c
#define IS_LEVEL_1       1
#define IS_LEVEL_2       2
#define IS_LEVEL_1_AND_2 3
```

`IS_LEVEL_1_AND_2` was conveniently set to `3` so that `(IS_LEVEL_1_AND_2 & IS_LEVEL_1)` and `(IS_LEVEL_1_AND_2 & IS_LEVEL_2)` will be different from zero.

rwestphal commented:
New update:
* Rebased the code;
* Added a 12th step in the new topotest to ensure isisd rejects invalid SR configuration.

rwestphal commented on isisd/isis_main.c:
```diff
@@ -193,13 +196,16 @@ FRR_DAEMON_INFO(isisd, ISIS, .vty_port = ISISD_VTY_PORT,
 int main(int argc, char **argv, char **envp)
 {
 	int opt;
+	int instance = 1;
```
Are you sure about this? The LM identifies the client daemons by both their protocol and instance ID. isisd and fabricd using the same instance number shouldn't be a problem.

rwestphal commented on isisd/isis_sr.c:
```diff
@@ -0,0 +1,1628 @@
+					    const struct sr_prefix_cfg *b)
+{
+	return prefix_cmp(&a->prefix, &b->prefix);
+}
```
The code only supports standard SPF, so I see no need to change this now. Also, we don't know yet if we'll want to put all Prefix-SIDs in the same tree or create separate trees for each supported SPF algorithm.

rwestphal commented on isisd/isis_sr.c:
```diff
@@ -0,0 +1,1628 @@
+
+	return 0;
+}
+
```
Just tested it here and it works normally.

rwestphal commented on isisd/isis_sr.c:
```diff
@@ -0,0 +1,1628 @@
+/* Parse all SR-related information from the given Router Capabilities TLV. */
+static struct sr_node *
+isis_sr_parse_route_cap_tlv(struct isis_area *area, int level,
+			    const uint8_t *sysid,
```
fixed the function name

rwestphal commented on isisd/isis_sr.c:
```diff
@@ -0,0 +1,1628 @@
+#else
+	srdb->config.enabled = false;
+	srdb->config.srgb_lower_bound = 16000;
+	srdb->config.srgb_upper_bound = 23999;
```
Fixed. I used these hard coded values because fabricd should be converted to YANG too in the short term, but that's probably not a good excuse :)

rwestphal commented on isisd/isis_sr.c:
```diff
@@ -0,0 +1,1628 @@
+	memset(srdb, 0, sizeof(*srdb));
+	srdb->enabled = false;
+	for (int level = ISIS_LEVEL1; level <= ISIS_LEVELS; level++) {
+		tree_sr_adjacency_init(&srdb->adj_sids);
```
fixed

rwestphal commented on tests/topotests/isis-sr-topo1/test_isis_sr_topo1.py:
```diff
@@ -0,0 +1,915 @@
+
+#
+# Step 1
+#
```
I think that's a good idea. I'll keep this in mind next time I update this topotest. Feel free to modify it as well if you want.

rwestphal commented:
> Renato,
> 
> There is major drawback regarding the Adjacency SID. As there are only a local significance, there must be linked to the SR-Node which advertise them to be usable.

Olivier,

Thanks again for your review.

The current code completely ignores remote Adj-SIDs since there's nothing we need to do with them for basic SR functionality.

For TI-LFA you're right that we'll need to keep track of Adj-SIDs on a per-node basis in order to compute the backup paths (remote Adj-SIDs should be needed to connect adjacent routers from the P-space and Q-space when a PQ node doesn't exist). But we'll need much more than just adding a back pointer to `sr_node` in the `sr_adjacency` structure. We'll also need to parse the remote Adj-SIDs, calculate and store the adjacent `sr_node` they point to, among a few other things. Local and remote Adj-SIDs will need to be handled differently the same way we handle local and remote Prefix-SID differently. That's a considerable amount of work to do, and I don't think this needs to be done as part of this PR. I'm already implementing TI-LFA on a separate branch and I plan to do this work there (handling of remote Adj-SIDs). Before that I'm focusing on the TI-LFA basic building blocks (Reverse SPF, computation of the P-space and Q-space and their intersection, microloop avoidance, etc).

Other than that, I think I addressed all of your other review comments. If I didn't please let me know.

odd22 commented:
Renato

> The current code completely ignores remote Adj-SIDs since there's nothing we need to do with them for basic SR functionality.

OK. I understand that the RB_TREE for ADJ-SID in the SRDB is only to manage local Adjacency SID. But, in this case, why not add it to the config structure within the SRDB? and replace the RB_TREE by a simple list (as there is no performance issue here)? In addition, since you are storing the ADJ-SID in isis_adjacency structure for local management, this is redundant. A choice must be made here.

When I requested to add ADJ-SID in the SRDB, it is for remote management of ADJ-SID (even if there are not yet parsed and stored) to prepare TI-LFA support. Not for local management of the ADJ-SID (even if it could also serve to that purpose).

More generally, I prefer that the data structures and more particularly the database schemas are designed by taking into account all the features that we want to implement. Rather than changing data structures when new features are introduced (which were known and planned at the beginning) because this can have an impact on the code already written. But of course, these are still individual design rules.

rwestphal commented:
> > The current code completely ignores remote Adj-SIDs since there's nothing we need to do with them for basic SR functionality.
> 
> OK. I understand that the RB_TREE for ADJ-SID in the SRDB is only to manage local Adjacency SID. But, in this case, why not add it to the config structure within the SRDB? and replace the RB_TREE by a simple list (as there is no performance issue here)? In addition, since you are storing the ADJ-SID in isis_adjacency structure for local management, this is redundant. A choice must be made here.

The `config` field of the `isis_sr_db` structure is reserved to hold configuration data coming from the northbound callbacks only. Regarding replacing the `sr_adjacency` rb-tree by a linked-list, I don't think that's a good idea. The TI-LFA algorithm is very CPU intensive, so I do think fast Adj-SID lookups will be necessary.

> When I requested to add ADJ-SID in the SRDB, it is for remote management of ADJ-SID (even if there are not yet parsed and stored) to prepare TI-LFA support. Not for local management of the ADJ-SID (even if it could also serve to that purpose).
> 
> More generally, I prefer that the data structures and more particularly the database schemas are designed by taking into account all the features that we want to implement. Rather than changing data structures when new features are introduced (which were known and planned at the beginning) because this can have an impact on the code already written. But of course, these are still individual design rules.

I could easily add a back pointer in the `sr_adjacency` structure like you did in your code, but my point is that this is largely insufficient.

This is how you declared `sr_adjacency` in your code:
```c
struct sr_adjacency {

        /* Adjacency & LAN Adjacency SID */
        struct isis_adj_sid *adj_sid;
        struct isis_lan_adj_sid *lan_sid;

        /* Input Label and nexthop for this Adjacency */
        mpls_label_t label;
        struct {
                int family;
                union g_addr ip;
                ifindex_t ifindex;
                mpls_label_t label;
        } nexthop;

        /* Back pointer to SR Node which advertise this Adjacency */
        struct sr_node *srn;

        /* Back pointer to isis adjacency to which SR Adjacency is attached */
        struct isis_adjacency *adj;
};
```

Most of these fields are only relevant to local Adj-SIDs (the Sub-TLV pointers, the whole `nexthop` structure and the `adj` back pointer). At the same time, there's missing information we'll need for remote Adj-SIDs, like the nexthop `sr_node`. What I plan to do is to create a `type` field in the `sr_adjacency` structure and add an union to separate which fields are only relevant to local Adj-SIDs and which fields are only relevant to remote Adj-SIDs. This is exactly what I did in the `sr_prefix` structure, and that allowed me to write code that is easy to understand. As I said before, I'm already implementing TI-LFA so if I were you I wouldn't worry too much about this since I'll have to do this work sooner or later.

rwestphal commented:
Closing...

As discussed in the weekly call yesterday, I'll work in cooperation with Olivier in a feature branch.
